nexusutils/dataconverter/readers/xps/reader.py

```diff
@@ -0,0 +1,349 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import os
+from pathlib import Path
+from typing import Any, Dict, Set
+from typing import Tuple
+import sys
+import json
+
+import yaml
+import numpy as np
+
+from nexusutils.dataconverter.readers.base.reader import BaseReader
+from nexusutils.dataconverter.readers.xps.reader_utils import XpsDataFileParser
+from nexusutils.dataconverter.readers.utils import flatten_and_replace, FlattenSettings
+from nexusutils.dataconverter.template import Template
+
+np.set_printoptions(threshold=sys.maxsize)
+
+XPS_TOCKEN = "@xps_tocken:"
+XPS_DATA_TOCKEN = "@data:"
+XPS_DETECTOR_TOCKEN = "@detector_data:"
+ELN_TOCKEN = "@eln"
+# Track entries for using for eln data
+ENTRY_SET: Set[str] = set()
+DETECTOR_SET: Set[str] = set()
+POSSIBLE_ENTRY_PATH: Dict = {}
+
+
+CONVERT_DICT = {
+    'Instrument': 'INSTRUMENT[instrument]',
+    'Analyser': 'ELECTRONANALYSER[electronanalyser]',
+    'Beam': 'BEAM[beam]',
+    'unit': '@units',
+    'version': '@version',
+    'Sample': 'SAMPLE[sample]',
+    'User': 'USER[user]',
+    'Data': 'DATA[data]',
+    'Source': 'SOURCE[source]',
+    'Collectioncolumn': 'COLLECTIONCOLUMN[collectioncolumn]',
+    'Energydispersion': 'ENERGYDISPERSION[energydispersion]',
+    'Detector': 'DETECTOR[detector]'
+}
+
+REPLACE_NESTED: Dict[str, str] = {}
+
+
+def construct_entry_name(key):
+    """Construct entry name from vendor, sample_name and region name"""
+
+    components = key.split("/")
+    try:
+        # entry: vendor__sample__name_of_scan_rerion
+        entry_name = (f'{components[2]}'
+                      f'__'
+                      f'{components[3].split("_", 1)[1]}'
+                      f'__'
+                      f'{components[5].split("_", 1)[1]}'
+                      )
+    except IndexError:
+        entry_name = ""
+    return entry_name
+
+
+def find_entry_and_value(xps_data_dict,
+                         key_part,
+                         dt_typ):
+
+    """Construct the entry name and pick up the corresponding data for
+       that for that entry.
+    """
+
+    entries_values = {}
+    if dt_typ == XPS_TOCKEN:
+
+        for key, val in xps_data_dict.items():
+            if key_part in key:
+                entry = construct_entry_name(key)
+                entries_values[entry] = val
+
+    elif dt_typ in (XPS_DATA_TOCKEN, XPS_DETECTOR_TOCKEN):
+        # entries_values = entry:{cycle0_scan0_chan0:xr.data}
+        entries_values = xps_data_dict["data"]
+
+    return entries_values
+
+
+# pylint: disable=too-many-locals
+def fill_data_group(key,
```

Comment:
Consider making these class functions as well. You won't need so many arguments and the silent pass by reference will not be needed.